File: solutions/4e49a099fd84fad956c856b0bb613db05b87316550730c4ac1329a2f26775465.py
# problem_id: 4e49a099fd84fad956c856b0bb613db05b87316550730c4ac1329a2f26775465
# source_language: rust
# phase: 3
# entrypoint: solve
# verified: True

def solve(stdin: str) -> str:
    import sys

    # Helper constants and functions for classification

    # Control: U+0000..U+001F and U+007F..U+009F
    def is_control(cp):
        return (0x0000 <= cp <= 0x001F) or (0x007F <= cp <= 0x009F)

    # Attachment ranges:
    # U+0300..U+036F, U+1AB0..U+1AFF, U+1DC0..U+1DFF,
    # U+20D0..U+20FF, U+FE00..U+FE0F, U+FE20..U+FE2F,
    # U+1F3FB..U+1F3FF, U+E0100..U+E01EF
    def is_attachment(cp):
        return (
            (0x0300 <= cp <= 0x036F) or
            (0x1AB0 <= cp <= 0x1AFF) or
            (0x1DC0 <= cp <= 0x1DFF) or
            (0x20D0 <= cp <= 0x20FF) or
            (0xFE00 <= cp <= 0xFE0F) or
            (0xFE20 <= cp <= 0xFE2F) or
            (0x1F3FB <= cp <= 0x1F3FF) or
            (0xE0100 <= cp <= 0xE01EF)
        )

    # Emoji ranges:
    # U+2600..U+27BF and U+1F000..U+1FAFF
    def is_emoji(cp):
        return (0x2600 <= cp <= 0x27BF) or (0x1F000 <= cp <= 0x1FAFF)

    # Regional Indicators: U+1F1E6..U+1F1FF
    def is_regional_indicator(cp):
        return 0x1F1E6 <= cp <= 0x1F1FF

    CR = 0x000D
    LF = 0x000A
    ZWJ = 0x200D

    tokens = stdin.split()
    if len(tokens) < 2:
        # No queries, return empty string
        return ''

    S_utf8 = tokens[0]
    Q = int(tokens[1])
    queries = tokens[2:]
    if len(queries) < 3 * Q:
        # Not enough queries, return empty string
        return ''

    # S_utf8 is the string S (already decoded)
    S = S_utf8
    scalars = [ord(c) for c in S]
    n = len(scalars)

    # Encode each scalar to UTF-8 bytes and track offsets
    utf8_bytes = []
    utf8_offsets = [0]
    utf16_units = []
    for cp in scalars:
        if cp <= 0x7F:
            b = bytes([cp])
        elif cp <= 0x7FF:
            b = bytes([
                0xC0 | (cp >> 6),
                0x80 | (cp & 0x3F)
            ])
        elif cp <= 0xFFFF:
            b = bytes([
                0xE0 | (cp >> 12),
                0x80 | ((cp >> 6) & 0x3F),
                0x80 | (cp & 0x3F)
            ])
        else:
            b = bytes([
                0xF0 | (cp >> 18),
                0x80 | ((cp >> 12) & 0x3F),
                0x80 | ((cp >> 6) & 0x3F),
                0x80 | (cp & 0x3F)
            ])
        utf8_bytes.append(b)
        utf8_offsets.append(utf8_offsets[-1] + len(b))
        utf16_units.append(1 if cp <= 0xFFFF else 2)

    # Precompute prev_non_attachment_idx for rule 4
    prev_non_attachment_idx = [-1] * n
    last_non_attachment = -1
    for i in range(n):
        if not is_attachment(scalars[i]):
            last_non_attachment = i
        prev_non_attachment_idx[i] = last_non_attachment

    # Precompute ri_run_len for rule 5
    ri_run_len = [0] * n
    for i in range(n):
        if is_regional_indicator(scalars[i]):
            ri_run_len[i] = 1 + (ri_run_len[i - 1] if i > 0 else 0)
        else:
            ri_run_len[i] = 0

    boundaries = [0]
    for i in range(n - 1):
        left = scalars[i]
        right = scalars[i + 1]

        # Rule 1: no boundary between CR and LF
        if left == CR and right == LF:
            # no boundary
            continue

        # Rule 2: boundary if either is control
        if is_control(left) or is_control(right):
            boundaries.append(i + 1)
            continue

        # Rule 3: no boundary if right is attachment or ZWJ
        if is_attachment(right) or right == ZWJ:
            continue

        # Rule 4: no boundary if left is ZWJ and right is emoji and nearest non-attachment before left is emoji
        if left == ZWJ and is_emoji(right):
            if i > 0:
                prev_idx = prev_non_attachment_idx[i - 1]
                if prev_idx != -1 and is_emoji(scalars[prev_idx]):
                    continue

        # Rule 5: no boundary if both are regional indicators and count of consecutive RIs ending at left is odd
        if is_regional_indicator(left) and is_regional_indicator(right):
            count = ri_run_len[i]
            if count % 2 == 1:
                continue

        # Rule 6: boundary
        boundaries.append(i + 1)

    boundaries.append(n)
    G = len(boundaries) - 1

    grapheme_utf8_offsets = []
    grapheme_utf16_units = []
    for gi in range(G):
        start_scalar = boundaries[gi]
        end_scalar = boundaries[gi + 1]
        start_byte = utf8_offsets[start_scalar]
        end_byte = utf8_offsets[end_scalar]
        grapheme_utf8_offsets.append((start_byte, end_byte))
        units = sum(utf16_units[start_scalar:end_scalar])
        grapheme_utf16_units.append(units)

    def resolve_index(x):
        if x < 0:
            x = G + x
        if x < 0:
            x = 0
        elif x > G:
            x = G
        return x

    base = 911
    mask64 = 0xFFFFFFFFFFFFFFFF

    def hash_bytes(bs):
        h = 0
        for b in bs:
            h = (h * base + b) & mask64
        return h

    grapheme_hashes = []
    for gi in range(G):
        start_scalar = boundaries[gi]
        end_scalar = boundaries[gi + 1]
        bs = b''.join(utf8_bytes[start_scalar:end_scalar])
        h = hash_bytes(bs)
        grapheme_hashes.append(h)

    prefix_hash = [0] * (G + 1)
    pow_base = [1] * (G + 1)
    for i in range(G):
        prefix_hash[i + 1] = (prefix_hash[i] * base + grapheme_hashes[i]) & mask64
        pow_base[i + 1] = (pow_base[i] * base) & mask64

    def get_hash(l, r):
        length = r - l
        h = prefix_hash[r] - ((prefix_hash[l] * pow_base[length]) & mask64)
        return h & mask64

    prefix_utf16 = [0] * (G + 1)
    for i in range(G):
        prefix_utf16[i + 1] = prefix_utf16[i] + grapheme_utf16_units[i]

    out = []
    idx = 0
    for _ in range(Q):
        t = int(queries[idx]); idx += 1
        a = int(queries[idx]); idx += 1
        b = int(queries[idx]); idx += 1
        i = resolve_index(a)
        j = resolve_index(b)
        if t == 1:
            if j >= i:
                out.append(str(prefix_utf16[i]))
                out.append(str(prefix_utf16[j]))
            else:
                out.append(str(prefix_utf16[i]))
                out.append(str(prefix_utf16[i]))
        else:
            maxL = min(G - i, G - j)
            low, high = 0, maxL
            while low < high:
                mid = (low + high + 1) // 2
                if get_hash(i, i + mid) == get_hash(j, j + mid):
                    low = mid
                else:
                    high = mid - 1
            L = low
            units = prefix_utf16[i + L] - prefix_utf16[i]
            out.append(str(L))
            out.append(str(units))

    return ' '.join(out)

File: solutions/4e49a099fd84fad956c856b0bb613db05b87316550730c4ac1329a2f26775465.rs
// problem_id: 4e49a099fd84fad956c856b0bb613db05b87316550730c4ac1329a2f26775465
// phase: 3
// entrypoint: main
// verified: True

use std::io::{self, Read};

const BASE: u64 = 911;

/// true if `b` is an ASCII whitespace (the token separator)
#[inline]
fn is_ws(b: u8) -> bool {
    matches!(b, 0x09..=0x0D | 0x20)
}

/// fast signed integer parser
fn parse_i64(data: &[u8], pos: &mut usize) -> i64 {
    while *pos < data.len() && is_ws(data[*pos]) {
        *pos += 1;
    }
    let mut sign = 1i64;
    if data[*pos] == b'-' {
        sign = -1;
        *pos += 1;
    }
    let mut val: i64 = 0;
    while *pos < data.len() && !is_ws(data[*pos]) {
        val = val * 10 + (data[*pos] - b'0') as i64;
        *pos += 1;
    }
    val * sign
}

/// classification helpers
#[inline]
fn is_control(cp: u32) -> bool {
    (0x0000..=0x001F).contains(&cp) || (0x007F..=0x009F).contains(&cp)
}
#[inline]
fn is_attachment(cp: u32) -> bool {
    (0x0300..=0x036F).contains(&cp)
        || (0x1AB0..=0x1AFF).contains(&cp)
        || (0x1DC0..=0x1DFF).contains(&cp)
        || (0x20D0..=0x20FF).contains(&cp)
        || (0xFE00..=0xFE0F).contains(&cp)
        || (0xFE20..=0xFE2F).contains(&cp)
        || (0x1F3FB..=0x1F3FF).contains(&cp)
        || (0xE0100..=0xE01EF).contains(&cp)
}
#[inline]
fn is_emoji(cp: u32) -> bool {
    (0x2600..=0x27BF).contains(&cp) || (0x1F000..=0x1FAFF).contains(&cp)
}
#[inline]
fn is_ri(cp: u32) -> bool {
    (0x1F1E6..=0x1F1FF).contains(&cp)
}

fn main() {
    // ----- read whole stdin -------------------------------------------------
    let mut input = Vec::new();
    io::stdin().read_to_end(&mut input).unwrap();

    // ----- first token = the UTF‑8 string S ---------------------------------
    let mut pos = 0usize;
    while pos < input.len() && is_ws(input[pos]) {
        pos += 1;
    }
    let start_s = pos;
    while pos < input.len() && !is_ws(input[pos]) {
        pos += 1;
    }
    let s_bytes = &input[start_s..pos]; // slice of the original bytes

    // ----- remaining tokens are numbers ------------------------------------
    let q: usize = parse_i64(&input, &mut pos) as usize;

    // ----- decode S into code points ----------------------------------------
    let s_str = std::str::from_utf8(s_bytes).expect("invalid UTF-8");
    let mut cps: Vec<u32> = Vec::with_capacity(s_str.len());
    let mut utf8_off: Vec<usize> = Vec::with_capacity(s_str.len() + 1);
    let mut utf16_pref: Vec<u64> = Vec::with_capacity(s_str.len() + 1);
    let mut is_att_vec: Vec<bool> = Vec::with_capacity(s_str.len());

    utf8_off.push(0);
    utf16_pref.push(0);
    for ch in s_str.chars() {
        let cp = ch as u32;
        cps.push(cp);
        let len_utf8 = ch.len_utf8();
        let len_utf16 = if cp <= 0xFFFF { 1u64 } else { 2u64 };
        let last_off = *utf8_off.last().unwrap();
        utf8_off.push(last_off + len_utf8);
        let last_utf16 = *utf16_pref.last().unwrap();
        utf16_pref.push(last_utf16 + len_utf16);
        is_att_vec.push(is_attachment(cp));
    }
    let n = cps.len();

    // ----- prev_non_attachment ------------------------------------------------
    let mut prev_non_att: Vec<usize> = vec![usize::MAX; n];
    let mut last_non = usize::MAX;
    for i in 0..n {
        if !is_att_vec[i] {
            last_non = i;
        }
        prev_non_att[i] = last_non;
    }

    // ----- ri run length ----------------------------------------------------
    let mut ri_run: Vec<usize> = vec![0; n];
    for i in 0..n {
        if is_ri(cps[i]) {
            ri_run[i] = 1 + if i > 0 { ri_run[i - 1] } else { 0 };
        }
    }

    // ----- compute grapheme boundaries ---------------------------------------
    let mut boundaries: Vec<usize> = Vec::with_capacity(n + 2);
    boundaries.push(0);
    const CR: u32 = 0x000D;
    const LF: u32 = 0x000A;
    const ZWJ: u32 = 0x200D;

    for i in 0..n - 1 {
        let left = cps[i];
        let right = cps[i + 1];

        // Rule 1
        if left == CR && right == LF {
            continue;
        }
        // Rule 2
        if is_control(left) || is_control(right) {
            boundaries.push(i + 1);
            continue;
        }
        // Rule 3
        if is_att_vec[i + 1] || right == ZWJ {
            continue;
        }
        // Rule 4
        if left == ZWJ && is_emoji(right) {
            if i > 0 {
                let prev = prev_non_att[i - 1];
                if prev != usize::MAX && is_emoji(cps[prev]) {
                    continue;
                }
            }
        }
        // Rule 5
        if is_ri(left) && is_ri(right) {
            let cnt = ri_run[i];
            if cnt % 2 == 1 {
                continue;
            }
        }
        // Rule 6
        boundaries.push(i + 1);
    }
    boundaries.push(n);
    let g = boundaries.len() - 1; // number of graphemes

    // ----- per‑grapheme data -------------------------------------------------
    let mut g_hash: Vec<u64> = Vec::with_capacity(g);
    let mut pref_utf16: Vec<u64> = Vec::with_capacity(g + 1);
    pref_utf16.push(0);

    for gi in 0..g {
        let start = boundaries[gi];
        let end = boundaries[gi + 1];
        // UTF‑16 length
        let utf16_len = utf16_pref[end] - utf16_pref[start];
        let next_utf16 = pref_utf16[gi] + utf16_len;
        pref_utf16.push(next_utf16);

        // hash of the UTF‑8 bytes of the grapheme
        let mut h: u64 = 0;
        let mut idx = utf8_off[start];
        while idx < utf8_off[end] {
            h = h.wrapping_mul(BASE).wrapping_add(s_bytes[idx] as u64);
            idx += 1;
        }
        g_hash.push(h);
    }

    // ----- prefix hash over grapheme hashes ----------------------------------
    let mut pref_hash: Vec<u64> = vec![0; g + 1];
    let mut pow_base: Vec<u64> = vec![1; g + 1];
    for i in 0..g {
        pref_hash[i + 1] = pref_hash[i]
            .wrapping_mul(BASE)
            .wrapping_add(g_hash[i]);
        pow_base[i + 1] = pow_base[i].wrapping_mul(BASE);
    }

    // ----- helper closures ---------------------------------------------------
    let resolve = |x: i64| -> usize {
        let mut idx = if x < 0 { (g as i64) + x } else { x };
        if idx < 0 {
            idx = 0;
        } else if idx > g as i64 {
            idx = g as i64;
        }
        idx as usize
    };
    let get_hash = |l: usize, r: usize,
                    pref: &Vec<u64>,
                    pow: &Vec<u64>|
     -> u64 {
        let len = r - l;
        let left = pref[l].wrapping_mul(pow[len]);
        pref[r].wrapping_sub(left)
    };

    // ----- answer queries ----------------------------------------------------
    let mut out = String::new();
    for _ in 0..q {
        let typ = parse_i64(&input, &mut pos) as i32;
        let a = parse_i64(&input, &mut pos);
        let b = parse_i64(&input, &mut pos);
        let i = resolve(a);
        let j = resolve(b);
        if typ == 1 {
            if j >= i {
                out.push_str(&format!("{} {}\n", pref_utf16[i], pref_utf16[j]));
            } else {
                out.push_str(&format!("{} {}\n", pref_utf16[i], pref_utf16[i]));
            }
        } else {
            // type 2
            let max_len = std::cmp::min(g - i, g - j);
            let mut lo = 0usize;
            let mut hi = max_len;
            while lo < hi {
                let mid = (lo + hi + 1) / 2;
                let h1 = get_hash(i, i + mid, &pref_hash, &pow_base);
                let h2 = get_hash(j, j + mid, &pref_hash, &pow_base);
                if h1 == h2 {
                    lo = mid;
                } else {
                    hi = mid - 1;
                }
            }
            let l = lo;
            let utf16_len = pref_utf16[i + l] - pref_utf16[i];
            out.push_str(&format!("{} {}\n", l, utf16_len));
        }
    }

    print!("{}", out);
}
